### DISCUSSION TITLE
Call for telemetry over mqtt.

### DISCUSSION THREAD

**Discussion by ssjophus** [arendst/Tasmota#15525] [Q&A] [answered]
Just a question that I think I already know the answer to, but here we go. I am using Tasmota to facilitate calibration of industrial instruments. I am using a few nodeMCU devices to read a relay state (as a switch). this triggers a state change in Tasmota and imediately sends an updated state over mqtt. This is fine. However, I plan to use the same setup to bring in 0-20mA data over I2C bus, and possibly data over serial. The problem I see is that for a calibration to be accurate, The data needs to come in at the exact right time. I don't need regular telemety updates, I need to be able to call for a telemery update at the right time. the 10 second minimum update time is to slow. 1 second updates would be better but Tasmota does not allow for that. 
Example: I can bring in data from a pressure sensor, I need to receive telemetry data when the transducer reaches an exact pressure. So Ideally Node Red will see the target pressure, then send a request for sensor data from the nodeMCU board measuring current. this way, the pressure and current come in within ms of each other for accurate measurements.
So, (with that background) Can I turn off regular telemetry updates, and instead, call for one on over mqtt command from Node Red?

**Comment by barbudor** [comment, marked as answer]
You can use a rule to follow the measure "real-time" and do the comparison on Tasmota:
Assuming your sensor is named XXX , provides you a pressure in bar and your target pressure is 3.0 bar 
```
Rule1 ON XXX#pressure>=3.0 DO publish some/topic/pressure_reached True ENDON
Rule1 1
Rule1 5
```
This will make Tasmota to publish on `some/topic/pressure_reached` the ayload `True` as soon as the pressure is equal or above 3.0
With "once" mode (Rule1 5), the publication will be done only once per crossing the limit (pressure will need to first drop below 3.0 to allow a new publication).

This will run at the fastest speed possible for Tasmota and depending on the sensor driver.

**Reply by sfromis** [reply]
If your idea of "when the transducer reaches an exact pressure" has a sub-second resolution, you're out of luck. Tasmota is not scoped for industrial automation with real-time data flows.

**Reply by ssjophus** [reply]
At the moment, My pressure transducer is a seperate unit. currently it has no connection to node red. (That is a seperate isue). 
If I can read serial data from my pressure transducer through a Tasmots device, would "barbudor's" solution be possable? 
I am also in full control of the rate of pressure increase. I can slow the increase of pressure to hang that reading for usualy no more than 2-3 seconds before the system drifts. As I said, I don't need regular intervals. Nor do I need constant real time status. I only need the value from the nodeMCU when the pressure transducer (seperate unit) reaches a set mark.
What I need is to to be able to call the nodeMCU and get current "stat/AI01/SENSOR" data.
I don't think there is a way to do it. However If I don't ask people who are smarter than me than the answer is garanteed to be no. If the answer is still no, than I may bring this up in the issues forum.

**Comment by barbudor** [comment]
You can still request a sensor message using `Status 10` but that's not necesseraly the best option. You will always get better responsivness by managing it directly on Tasmota, given you have a driver for your sensor

So you have a pressure sensor that outputs serial
- Is there an existing driver for Tasmota ? 
- Or are you planning to write one ?
- If not where do you expect the "translation" from the serial data into pressure information to be performed ?

Have you considered the following possible alternatives :
- Writing a native Tasmota driver for your sensor (in C)
- Writing a Tasmota Berry driver for your sensor (requires an ESP32)
- Using TCPBridge to transfer the serial stream over Wifi from the Tasmota device to a Node-RED tcp node and process the serial data in Node-RED

**Reply by ssjophus** [reply]
I don't know if a driver exists for my pressure transducer (Fluke P700), I am guessing no. I don't know the first thing about writing a driver. Fluke is NOT at ALL forthcomming about outside devices communicating with any of there stuff. I am not apposed to using an esp32 if that's what it takes get it to work. I know someone that can figure out the fluke end of things without bricking it. I have to use the Fluke transducer because TransCat can certify its calibration

**Reply by ssjophus** [reply]
I think the only way to get around the problem is to learn Arduino coding and build a sketch that updates my sensor data every scan, or delayed to 500 or so ms to regulate data transfer. Or use an mqtt capable PLC to bring in real time data (expensive).

**Reply by barbudor** [reply]
Using Tasmota and Berry scripting language (close to Python) on an ESP32 will allow to take advantage of an existing framework and get, IMHO, something more maintainable than an Arduino sketch

**Reply by ssjophus** [reply]
I will look into that.
Thank you,
